src/routes/tradesToPublish.ts

```diff
@@ -0,0 +1,102 @@
+import { Request } from "express";
+import { Response } from "express";
+import { getLogger } from "log4js";
+import { ComitMetadata } from "../comitNode";
+import { BitcoinConfig } from "../config";
+import Ledger from "../ledger";
+import { Trade, TradeService } from "../rates/tradeService";
+import { EthereumWallet } from "../wallets/ethereum";
+
+const logger = getLogger();
+
+const previousTradesLookup: Map<string, Trade> = new Map<string, Trade>();
+
+export function getAmountsToPublishRoute(
+  tradeService: TradeService,
+  bitcoinConfig: BitcoinConfig,
+  ethereumWallet: EthereumWallet,
+  comitMetadata: ComitMetadata
+) {
+  // @ts-ignore
+  return async (req: Request, res: Response) => {
+    try {
+      if (!bitcoinConfig) {
+        throw new Error("Bitcoin not configured correctly.");
+      }
+      if (!ethereumWallet) {
+        throw new Error("Ethereum not configured correctly");
+      }
+
+      const trades = await tradeService.prepareTradesToPublish();
+      const addressHint =
+        comitMetadata.listenAddresses.length > 0
+          ? comitMetadata.listenAddresses[0]
+          : undefined;
+
+      const publishTradesResponse = {
+        peerId: comitMetadata.id,
+        addressHint,
+        ledgers: [
+          {
+            name: Ledger.Bitcoin,
+            network: bitcoinConfig.network
+          },
+          {
+            name: Ledger.Ethereum,
+            network: ethereumWallet.getChainId() === 17 ? "regtest" : "testnet" // TODO: refactor Ethereum ChainId handling in bobtimus
+          }
+        ],
+        rates: trades.map(trade => {
```

Comment:
❓ why not storing the timestamp in the trade?
This logic seems very convoluted and bug-prone. What are the design reasons for it?

Reply:
Mhm.. I did not touch this. 
What do you mean exactly and how would you do it differently?

Reply:
What I mean is that here we store the timestamp in a hashmap where the key is the trade amounts.

if no value is found, we add the timestamp to the hashmap.

I would expect to store the full trade (including the timestamp) in a hashmap as value. The key would be the trading pair. Only one trade would be stored for a given key.

With the current code, the following would happen for buy eth, sell btc offer (why is it call trade? it's an offer until it becomes a trade):
1. Rate is 1000 ether for 1 btc
2. API request, no timestamp is available, create timestamp 1-jan-19
3. Rate becomes 900 ether for 1 btc
4. API request, no timestamp is available, create timestamp 2-jan-19 (memory can increase in an unlimited manner)
5. Rate becomes 1000 ether for 1 btc
6.  API request, timestamp is available! returns **1-jan-19** <- this is not good

Reply:
I think you are misreading something or you reviewed an old version?

> What I mean is that here we store the timestamp in a hashmap where the key is the trade amounts.

The value in the map is the `trade`:
https://github.com/coblox/bobtimus/blob/39d24b5faa81310d53ac012809574329237f8c2d/src/routes/tradesToPublish.ts#L12
The key of the value is the ledger-asset combination: 
https://github.com/coblox/bobtimus/blob/39d24b5faa81310d53ac012809574329237f8c2d/src/routes/tradesToPublish.ts#L77

> I would expect to store the full trade (including the timestamp) in a hashmap as value. The key would be the trading pair. Only one trade would be stored for a given key.

that's what's happening:

https://github.com/coblox/bobtimus/blob/39d24b5faa81310d53ac012809574329237f8c2d/src/routes/tradesToPublish.ts#L85

Tests are a bit hidden though, I'll add some better onces.